File: firmware/01-v1/master/power_manager.py
"""
GridBox — Power Manager
ADC-based power monitoring: bus voltage, motor currents, power calculations.
"""

from machine import Pin, ADC
import config


class PowerManager:
    """Reads ADC channels and computes voltage, current, power, efficiency."""

    def __init__(self, adc_bus_pin=None, adc_m1_pin=None, adc_m2_pin=None,
                 r_sense=None, divider_ratio=None):
        self.adc_bus = ADC(Pin(adc_bus_pin or config.ADC_BUS_VOLTAGE))
        self.adc_m1 = ADC(Pin(adc_m1_pin or config.ADC_MOTOR1_CURRENT))
        self.adc_m2 = ADC(Pin(adc_m2_pin or config.ADC_MOTOR2_CURRENT))
        self.r_sense = r_sense or config.CURRENT_SENSE_R
        self.divider_ratio = divider_ratio or config.VOLTAGE_DIVIDER_RATIO
        self.n_samples = config.ADC_SAMPLES_AVG

        # Tracking
        self._total_energy_mWs = 0.0
        self._useful_energy_mWs = 0.0
        self._last_read_ms = 0

    def _read_adc_avg(self, adc):
        """Read ADC with N-sample averaging for noise reduction."""
        total = 0
        for _ in range(self.n_samples):
            total += adc.read_u16()
        return total / self.n_samples

    def _adc_to_voltage(self, raw):
        """Convert raw ADC reading to voltage."""
        return raw * config.ADC_VREF / config.ADC_RESOLUTION

    def read_bus_voltage(self):
        """Read bus voltage in V (with divider correction)."""
        raw = self._read_adc_avg(self.adc_bus)
        return self._adc_to_voltage(raw) * self.divider_ratio

    def read_motor_current(self, motor_id):
        """Read motor current in mA.

        Args:
            motor_id: 1 or 2
        """
        adc = self.adc_m1 if motor_id == 1 else self.adc_m2
        raw = self._read_adc_avg(adc)
        v_sense = self._adc_to_voltage(raw)
        return (v_sense / self.r_sense) * 1000  # mA

    def read_all(self):
        """Read all channels and compute power metrics.

        Returns dict:
            bus_v: bus voltage (V)
            m1_mA: motor 1 current (mA)
            m2_mA: motor 2 current (mA)
            m1_W: motor 1 power (W)
            m2_W: motor 2 power (W)
            total_W: total power (W)
            excess_W: available excess power (W)
            efficiency: useful/total percentage
        """
        bus_v = self.read_bus_voltage()
        m1_mA = self.read_motor_current(1)
        m2_mA = self.read_motor_current(2)

        m1_W = bus_v * m1_mA / 1000
        m2_W = bus_v * m2_mA / 1000
        total_W = m1_W + m2_W

        # Excess = what's available vs nominal capacity
        nominal_W = bus_v * config.MOTOR_CURRENT_MAX_MA / 1000 * 2
        excess_W = max(0, nominal_W - total_W) if nominal_W > 0 else 0

        # Efficiency: assume useful = actual, total = nominal
        efficiency = (total_W / nominal_W * 100) if nominal_W > 0 else 0

        return {
            'bus_v': round(bus_v, 2),
            'm1_mA': round(m1_mA, 1),
            'm2_mA': round(m2_mA, 1),
            'm1_W': round(m1_W, 3),
            'm2_W': round(m2_W, 3),
            'total_W': round(total_W, 3),
            'excess_W': round(excess_W, 3),
            'efficiency': round(efficiency, 1),
        }

    def get_efficiency(self):
        """Return power efficiency percentage."""
        data = self.read_all()
        return data['efficiency']

    def is_overloaded(self):
        """Return True if bus voltage is below critical threshold."""
        return self.read_bus_voltage() < config.BUS_VOLTAGE_CRITICAL

    def is_low_voltage(self):
        """Return True if bus voltage is below warning threshold."""
        return self.read_bus_voltage() < config.BUS_VOLTAGE_LOW

    def get_excess_power(self):
        """Return watts available for rerouting."""
        data = self.read_all()
        return data['excess_W']


if __name__ == "__main__":
    import time

    print("=" * 40)
    print("  Power Manager Test")
    print("=" * 40)

    pm = PowerManager()

    while True:
        data = pm.read_all()
        overloaded = pm.is_overloaded()
        low = pm.is_low_voltage()

        status = "OVERLOADED!" if overloaded else ("LOW" if low else "OK")
        print(f"Bus={data['bus_v']:.2f}V [{status}] | "
              f"M1={data['m1_mA']:.0f}mA ({data['m1_W']:.2f}W) | "
              f"M2={data['m2_mA']:.0f}mA ({data['m2_W']:.2f}W) | "
              f"Total={data['total_W']:.2f}W | "
              f"Excess={data['excess_W']:.2f}W | "
              f"Eff={data['efficiency']:.0f}%")

        time.sleep_ms(200)

File: src/master-pico/c_sdk/power_manager.c
/**
 * GridBox — Power Manager (C SDK)
 * ADC-based power monitoring: bus voltage, motor currents, power calculations.
 * Ported from MicroPython power_manager.py.
 */

#include "power_manager.h"
#include "hardware/adc.h"
#include <stdio.h>

/* ---- Internal helpers ---- */

static float read_adc_avg(power_manager_t *pm, uint8_t adc_input) {
    adc_select_input(adc_input);
    uint32_t total = 0;
    for (int i = 0; i < pm->n_samples; i++) {
        total += adc_read();
    }
    return (float)total / (float)pm->n_samples;
}

static float adc_to_voltage(float raw) {
    /* RP2350 ADC: 12-bit (0-4095), VREF = 3.3V */
    return raw * PM_ADC_VREF / 4095.0f;
}

/* ---- Public API ---- */

void power_manager_init(power_manager_t *pm, float r_sense,
                        float divider_ratio, uint8_t n_samples) {
    pm->r_sense = r_sense;
    pm->divider_ratio = divider_ratio;
    pm->n_samples = n_samples;

    /* Init ADC hardware */
    adc_init();
    adc_gpio_init(26);  /* GP26 = ADC0 = bus voltage */
    adc_gpio_init(27);  /* GP27 = ADC1 = motor 1 current */
    adc_gpio_init(28);  /* GP28 = ADC2 = motor 2 current */

    printf("[POWER] Init OK: Rsense=%.1fΩ, divider=%.1f, samples=%d\n",
           r_sense, divider_ratio, n_samples);
}

float power_manager_read_bus_voltage(power_manager_t *pm) {
    float raw = read_adc_avg(pm, PM_ADC_BUS_VOLTAGE);
    return adc_to_voltage(raw) * pm->divider_ratio;
}

float power_manager_read_motor_current(power_manager_t *pm, uint8_t motor_id) {
    uint8_t input = (motor_id == 1) ? PM_ADC_MOTOR1_CURRENT : PM_ADC_MOTOR2_CURRENT;
    float raw = read_adc_avg(pm, input);
    float v_sense = adc_to_voltage(raw);
    return (v_sense / pm->r_sense) * 1000.0f;  /* convert to mA */
}

power_reading_t power_manager_read_all(power_manager_t *pm) {
    power_reading_t r;

    r.bus_v = power_manager_read_bus_voltage(pm);
    r.m1_mA = power_manager_read_motor_current(pm, 1);
    r.m2_mA = power_manager_read_motor_current(pm, 2);

    /* Power = V * I */
    r.m1_W = r.bus_v * r.m1_mA / 1000.0f;
    r.m2_W = r.bus_v * r.m2_mA / 1000.0f;
    r.total_W = r.m1_W + r.m2_W;

    /* Nominal max power (both motors at max rated current) */
    float nominal_W = r.bus_v * PM_MOTOR_CURRENT_MAX_MA / 1000.0f * 2.0f;
    r.excess_W = (nominal_W > r.total_W) ? (nominal_W - r.total_W) : 0.0f;
    r.efficiency = (nominal_W > 0.0f) ? (r.total_W / nominal_W * 100.0f) : 0.0f;

    return r;
}

float power_manager_get_efficiency(power_manager_t *pm) {
    power_reading_t r = power_manager_read_all(pm);
    return r.efficiency;
}

bool power_manager_is_overloaded(power_manager_t *pm) {
    return power_manager_read_bus_voltage(pm) < PM_BUS_VOLTAGE_CRITICAL;
}

bool power_manager_is_low_voltage(power_manager_t *pm) {
    return power_manager_read_bus_voltage(pm) < PM_BUS_VOLTAGE_LOW;
}

float power_manager_get_excess_power(power_manager_t *pm) {
    power_reading_t r = power_manager_read_all(pm);
    return r.excess_W;
}
